R/unitConverter.R

```diff
@@ -0,0 +1,103 @@
+#'   not specified (`NULL`), first of the existing units in the respective
+#'   columns (`xUnit` and `yUnit`) will be selected as the common unit. For
+#'   available dimensions and units, see `ospsuite::ospDimensions` and
+#'   `ospsuite::ospUnits`, respectively.
+#'
+#' @seealso toUnit
+#'
+#' @examples
+#'
+#' # small dataframe to illustrate the conversion
+#' (df <- dplyr::tibble(
+#'   xValues = c(15, 30, 60), xUnit = "min", xDimension = "Time",
+#'   yValues = c(0.25, 45, 78), yUnit = c("", "%", "%"), yDimension = c("Fraction", "Fraction", "Fraction"),
+#'   molWeight = c(12.5, 10, 5)
+#' ))
+#'
+#' ospsuite:::.unitConverter(df)
+#' ospsuite:::.unitConverter(df, xUnit = ospUnits$Time$h)
+#' ospsuite:::.unitConverter(df, yUnit = ospUnits$Fraction$`%`)
+#' ospsuite:::.unitConverter(df, xUnit = ospUnits$Time$h, yUnit = ospUnits$Fraction$`%`)
+#'
+#' @keywords internal
+#' @noRd
+.unitConverter <- function(data, xUnit = NULL, yUnit = NULL) {
+  # No validation of inputs for this non-exported function.
+  # All validation will take place in the `DataCombined` class itself.
+
+  # The observed and simulated data should have same units.
+  #
+  # Therefore, if target units are not specified, we need to choose one for
+  # consistency. For no special reason, first element from a vector of unique
+  # units will be selected.
+  targetXUnit <- xUnit %||% .selectDefaultTargetUnit(data$xUnit)
+  targetYUnit <- yUnit %||% .selectDefaultTargetUnit(data$yUnit)
+
+  # *WARNING*: Do not change the order of two `mutate()` statements.
+  #
+  # Since the old `xUnit` and `yUnit` columns are need for unit conversion,
+  # those columns must be updated only *after* the conversions have taken place.
+  data %>%
+    dplyr::rowwise() %>% # transform values to common units
+    dplyr::mutate(
+      xValues = toUnit(
+        quantityOrDimension = xDimension,
+        values              = xValues,
+        targetUnit          = targetXUnit,
+        sourceUnit          = xUnit
+      ),
+      yValues = toUnit(
+        quantityOrDimension = yDimension,
+        values              = yValues,
+        targetUnit          = targetYUnit,
+        sourceUnit          = yUnit,
+        molWeight           = molWeight,
+        molWeightUnit       = ospUnits$`Molecular weight`$`g/mol`
+      )
+    ) %>% # update the columns with chosen target units
+    dplyr::mutate(
+      xUnit = targetXUnit,
+      yUnit = targetYUnit
+    ) %>%
+    dplyr::ungroup()
+}
+
+# helper -----------------------
+
+#' Select target unit
+#'
+#' @description
+#'
+#' Select appropriate target unit from the existing source units if no target
+#' unit was specified by the user.
+#'
+#' @keywords internal
+#' @noRd
+.selectDefaultTargetUnit <- function(sourceUnit) {
+  argName <- deparse(substitute(sourceUnit))
+  sourceUnitLevels <- unique(sourceUnit)
+
+  # `toUnit` can't carry out conversion if source unit is `NA`
+  if (any(is.na(sourceUnitLevels))) {
```

Comment:
This should NEVER happen, this implies that something went terrbly  wrong. As the whole purpose of this function is to check for `NA`, I would rather use `sourceUnitLevels[[1]]` in the parent function and not define a new function for it.